diff --git a/asbestos/gadgets-aarch64/control.S b/asbestos/gadgets-aarch64/control.S
index 6fad18ae..8202fbe4 100644
--- a/asbestos/gadgets-aarch64/control.S
+++ b/asbestos/gadgets-aarch64/control.S
@@ -94,9 +94,11 @@
 .endm
 
 .macro check_res_or_flag resflag, flag, target, fallthrough
-    ldr w8, [_cpu, CPU_flags_res]
+    add x9, _cpu, #CPU_flags_res
+ldr w8, [x9]
     tbnz w8, \resflag, 2f
-    ldr w8, [_cpu, CPU_eflags]
+    add x9, _cpu, #CPU_eflags
+ldr w8, [x9]
     tbnz w8, \flag, \target
     b \fallthrough
 2:
@@ -112,26 +114,30 @@
         cbnz w8, \target
     .else N .ifc \cond,z
         check_res_or_flag 1/*ZF_RES*/, 6/*ZF_FLAG*/, \target, 3f
-        ldr w8, [_cpu, CPU_res]
+        add x9, _cpu, #CPU_res
+ldr w8, [x9]
         cbz w8, \target
     3:
     .else N .ifc \cond,cz
         ldrb w8, [_cpu, CPU_cf]
         cbnz w8, \target
         check_res_or_flag 1/*ZF_RES*/, 6/*ZF_FLAG*/, \target, 3f
-        ldr w8, [_cpu, CPU_res]
+        add x9, _cpu, #CPU_res
+ldr w8, [x9]
         cbz w8, \target
     3:
     .else N .ifc \cond,s
         check_res_or_flag 2/*SF_RES*/, 7/*SF_FLAG*/, \target, 3f
-        ldr w8, [_cpu, CPU_res]
+        add x9, _cpu, #CPU_res
+ldr w8, [x9]
         cmp w8, 0
         b.lt \target
     3:
     .else N .ifc \cond,p
         check_res_or_flag 0/*PF_RES*/, 2/*PF_FLAG*/, \target, 3f
         # this is so sad
-        ldr w8, [_cpu, CPU_res]
+        add x9, _cpu, #CPU_res
+ldr w8, [x9]
         uxtb w8, w8
         fmov s0, w8
         cnt v0.8b, v0.8b
@@ -140,14 +146,16 @@
         tbz w8, 0, \target
     3:
     .else N .ifc \cond,sxo
-        ldr w8, [_cpu, CPU_res]
+        add x9, _cpu, #CPU_res
+ldr w8, [x9]
         cmp w8, 0
         cset w8, lt
         ldrb w9, [_cpu, CPU_of]
         cmp w8, w9
         b.ne \target
     .else N .ifc \cond,sxoz
-        ldr w8, [_cpu, CPU_res]
+        add x9, _cpu, #CPU_res
+ldr w8, [x9]
         cmp w8, 0
         b.eq \target
         cset w8, lt
@@ -205,7 +213,8 @@
     sub esp, esp, 4
     mov _addr, esp
     write_prep 32, pushf
-    ldr w8, [_cpu, CPU_eflags]
+    add x9, _cpu, #CPU_eflags
+ldr w8, [x9]
     str w8, [_xaddr]
     write_done 32, pushf
     gret
@@ -215,7 +224,8 @@
     mov _addr, esp
     read_prep 32, popf
     ldr w8, [_xaddr]
-    str w8, [_cpu, CPU_eflags]
+    add x9, _cpu, #CPU_eflags
+str w8, [x9]
     add esp, esp, 4
 
     save_c
@@ -227,7 +237,8 @@
 
 .gadget sahf
     ubfx w8, eax, 8, 8
-    strb w8, [_cpu, CPU_eflags]
+    add x9, _cpu, #CPU_eflags
+strb w8, [x9]
     save_c
     mov x0, _cpu
     bl NAME(helper_expand_flags)
